### ISSUE TITLE
`Vuex.Store` should also take in `actions` and `getters`

### ISSUE THREAD

**Issue by analog-nico** [vuejs/vuex#221]
We all know that `actions` and `getters` can be nicely injected into Vue components:

``` js
new Vue({ vuex: { actions: { ... }, getters: { ... } } })
```

That means you can easily use them within the context of a Vue instance.

And that is good because commonly used and shared `actions` and `getters` nicely improve the app's maintainability.

However, code that is outside the context of a Vue instance must take extra measures to properly call them. Calling `actions` is quite easy. Calling `getters` is more difficult. Their reactivity needs to be preserved by passing the ES5 getter instead of accidentally invoking it too early and just passing its returned value!

When using `vuex` together with `vue-router` I came across a case – https://github.com/vuejs/vue-router/issues/533 – were I need to access the application state inside the router's `canActivate` hook. When the hook is called the Vue instance does not exist. Hence, those extra measures would be needed to call the getter.

It is not a huge deal to handle that but why does the `Vuex.Store` only take in the `state` and `mutations`? IMHO the store could as well take in `actions` and `getters` so they can be called in the same very easy way everywhere in the codebase.

``` js
// store.js
export default new Vuex.Store({
    state: {
        count: 0
    },
    mutations: { ... },
    actions: { ... },
    getters: {
        getCount: (state) => state.count
    }
})
```

``` js
// somewhere-outside-a-vue-instance.js

import store from 'path/to/store.js'

store.getCount // :)
```

**Comment by fnlctrl** [comment]
I think the problem is already addressed in #157 :D, though the author didn't have time to implement it since he was very busy with vue 2.0.

**Comment by analog-nico** [comment]
Thanks @fnlctrl ! The issue is a great enhancement. After reading the comments multiple times I think it is about separating actions and getters into modules so that you can access them within the context of a Vue instance in a cleaner way. Did I miss something?

My main point here is – as you said in https://github.com/vuejs/vue-router/issues/533 – that I wish that the

> vuex store can be standalone (having its own getters), and we can directly access states via its getters, without the presence of vue instances

I couldn't have said it better! :)

**Comment by yyx990803** [comment]
What prevents you from doing `getter(store.state)`?

**Comment by analog-nico** [comment]
@yyx990803 Nothing, really. I am just wondering: If it is recommended to use getters instead of directly accessing the state why are getters separated from the store in the first place?

**Comment by yyx990803** [comment]
@analog-nico this allows child components to be completely de-coupled from the store singleton (it only needs to be injected at the root entry point once). If getters are coupled to the store, then you will need a reference to the store any time you use a getter.

**Comment by analog-nico** [comment]
Good points @yyx990803 ! I am not experienced enough to actually evaluate the effects of my suggestion. So I explain you my simplest use case now and you may just close this issue if you want.

I use `store.state.waiting = <boolean>` and `action.startWaiting()`, `action.stopWaiting()`, and `getter.isWaiting` throughout my app. It blocks/unblocks my app and shows a spinner etc.

I need to use those actions and getters:
1. outside the context of a Vue component – in the routers `canActivate` hook (see https://github.com/vuejs/vue-router/issues/533)
2. inside the context of common components that I share across multiple SPAs
3. inside the context of basically every custom component in my SPA

For 1. there are at least the two solutions discussed in https://github.com/vuejs/vue-router/issues/533 that keep the code maintainable. So this suggestion would be pure convenience.

For 2. I duplicated the `getter.isWaiting` but I could as well have created a common store and inherited SPA-specific stores plus actions and getters.

For 3. we have the `vuex` binding of course.

What would make my coding easier is if I could bind some actions and getters directly to the store and those are also injected into the Vue component which I bind to the store as well as propagated to the children like the store itself. Attention, I sneaked in the suggestion to propagate actions and getters down the component tree like the store. ;)

Altogether it would look like this:

``` js
// store.js
export default new Vuex.Store({
    state: {
        waiting: false
    },
    mutations: {
        SET_WAITING(state, value) {
            state.waiting = value
        }
    },
    actions: {
        startWaiting({ dispatch }) {
            dispatch('SET_WAITING', true)
        },
        stopWaiting({ dispatch }) {
            dispatch('SET_WAITING', false)
        }
    },
    getters: {
        isWaiting(state) {
            return state.waiting
        }
    }
})
```

``` js
// somewhere-outside-a-vue-instance.js

import store from 'path/to/store.js'

store.startWaiting()
```

``` js
// boot-app.js

import store from 'path/to/store.js'

let App = Vue.extend({
    store,
    vuex: {
        // startWaiting, stopWaiting, isWaiting are already bound
        // Extra actions and getters may be defined here
    }
})
```

``` js
// common-app-spinner-component-shared-across-multiple-spas.js

Vue.component('app-spinner', {
    vuex: {
        getters: {
            // Reimplementation can be dropped if the actions and getters are propagated down from the app
            // isWaiting(state) {
            //     return state.waiting
            // }
        }
    }
})
```

``` js
// custom-component.js

Vue.component('some-component', {
    vuex: {
        // startWaiting, stopWaiting, isWaiting are already bound due to propagation with the store
        // Extra actions and getters may be defined here
    }
})
```

**Comment by fnlctrl** [comment]
> What prevents you from doing getter(store.state)?
@yyx990803 Sorry if this question sounds stupid but, is the getter like this? 
```javascript
var getter = obj => obj
```

**Comment by yyx990803** [comment]
So you basically want globally bound actions and getters which are implicitly shared across all components - which I think is a bad idea. Also, this doesn't seem to address the problem of accessing state in `canActivate`.

Getters and actions do not "rely on a Vue instance" - rather, they rely on a store instance. When they are used in a Vue instance, they are simply using `this.$store` injected into that instance. So using them outside of a Vue instance is simply importing and then calling them directly with the store as an argument. Remember that they are just normal functions.

A lot of this comes down to whether you are ok with coupling your components to the store instance. If you are ok with that, then you can always just do `import store from './store'` and directly retrieve state or dispatch actions on it.

**Comment by fnlctrl** [comment]
So.. what is your suggestion about accessing states outside a vue instance, like in `canActivate`?
Even if we do `getter(store.state)`, we are still coupling the component to the store instance.

**Comment by yyx990803** [comment]
Resolved in 2.0.

**Comment by analog-nico** [comment]
Thank you very much! I read the v2.0.0-rc.1 release notes and really like all the changes.
